tools/libfrencutils/create_xgrid.c

```diff
@@ -2100,13 +2099,57 @@ double poly_ctrlat(const double x[], const double y[], int n)
     avg_y = (lat1+lat2)*0.5;
     if      (dx==0.0) continue;
     if(dx > M_PI)  dx = dx - 2.0*M_PI;
-    if(dx < -M_PI) dx = dx + 2.0*M_PI;
+    if(dx <= -M_PI) dx = dx + 2.0*M_PI; // flip sign for dx=-pi to fix huge value see comments in function poly_area
 
     if ( fabs(hdy)< SMALL_VALUE ) /* cheap area calculation along latitude */
       ctrlat -= dx*(2*cos(avg_y) + lat2*sin(avg_y) - cos(lat1) );
     else
       ctrlat -= dx*( (sin(hdy)/hdy)*(2*cos(avg_y) + lat2*sin(avg_y)) - cos(lat1) );
```

Comment:
This formula for is a black box to me. All I know at this point is that it supposedly produces (centroid latitude) * (area)  , some kind of moment of inertia for the polygon on the sphere. It is later divided by the area in fregrid before being used as the centroid and hence has to have the same fix for sides passing thru poles as in poly_area().  I guess the derivation is along the lines I sketched for poly_area, but I have not been able to derive it or find any reference for it yet. If you know a reference or the derivation please add it here.

Reply:
Nevermind, I found the derivation and the reference. It's Phil Jones's 1998 again. I'll add comments to the code.